### PR TITLE
Add Tagger

### PR DESCRIPTION
**Pull request by AmalieDue**
### Description
Added macor and macor_test to crypto folder in library. The Macor struct is empty which seems strange, but I am not sure what to insert. Some feedback would be helpful before I proceed :) 

closes DEV-81

### Type(s) of Change (Split the PR if you check many)
- [ ] Added user feature
- [x] Added technical feature
- [ ] Added tests
- [ ] Refactor
- [ ] Bugfix
- [ ] Updated documentation
- [ ] Kubernetes changes
- [ ] CI/CD workflow changes

### Testing Checklist
- [x] I have added/updated unit tests for functional changes.
- [ ] I have added/updated integration tests for changes that affect dependent systems.
- [ ] I have added/updated end-to-end tests for feature changes.

### General Checklist
- [x] I have pulled in the latest changes from master.
- [ ] I have run `make lint`.
- [ ] I have successfully run `make tests`.
- [ ] I have updated relevant CI/CD workflows.
- [ ] I have updated relevant Kubernetes manifests/scripts.
- [ ] I have updated/added relevant (internal and external) documentation (readme, doc comments, etc).
- [ ] I have updated the dependency map to reflect my changes.
- [ ] I have added the license to new source code files.
- [ ] I have spent some time looking over the full diff before creating this PR.

### CODE REVIEW

File: lib/crypto/interfaces.go [hunk 1]
```diff
@@ -26,6 +26,12 @@ type CryptorInterface interface {
 	Decrypt(plaintext, data interface{}, wrappedKey, ciphertext []byte) (err error)
 }
 
+// MacorInterface provides an API to create a MAC.
+type MacorInterface interface {
```
Comment (philip-cc) [inline lib/crypto/interfaces.go:30]: I'd suggest `Tagger` instead, as in "cryptographic tag".

File: lib/crypto/interfaces.go [hunk 2]
```diff
@@ -26,6 +26,12 @@ type CryptorInterface interface {
...
+	// MAC creates a MAC from a key, tag size, and optional customization string.
+	MAC(key []byte, tagSize int, customizationString interface{}) (mac []byte, err error)
```
Comment (philip-cc) [inline lib/crypto/interfaces.go:32]: `key` should not be part of the interface. Instead a key should be passed to the constructor of the implementing struct (see e.g. `NewAESCryptor`).

Comment (philip-cc) [inline lib/crypto/interfaces.go:32]: `tagSize` should be part of the implementation (compare with `EncryptionKeyLength`).

Comment (philip-cc) [inline lib/crypto/interfaces.go:32]: `customizationString` is a weird name for something that's not a string. We should probably follow the convention in `CryptorInterface` and call it `data`.

File: lib/crypto/interfaces.go (hunk 1 above)
Comment (AmalieDue) [reply lib/crypto/interfaces.go]: Fixed

File: lib/crypto/interfaces.go (hunk 2 above)
Comment (AmalieDue) [reply lib/crypto/interfaces.go]: Fixed

Comment (AmalieDue) [reply lib/crypto/interfaces.go]: Fixed

Comment (AmalieDue) [reply lib/crypto/interfaces.go]: Fixed

File: lib/crypto/interfaces.go [hunk 3]
```diff
@@ -26,6 +26,12 @@ type CryptorInterface interface {
 	Decrypt(plaintext, data interface{}, wrappedKey, ciphertext []byte) (err error)
 }
 
+// TaggerInterface provides an API to create a tag, i.e. a piece of information used for authenticating data.
+type TaggerInterface interface {
+	// Tag creates a tag of data.
```
Comment (philip-cc) [inline lib/crypto/interfaces.go:31]:
```suggestion
	// Tag creates a cryptographic tag of data.
```

File: lib/crypto/interfaces.go [hunk 4]
```diff
@@ -26,6 +26,12 @@ type CryptorInterface interface {
 	Decrypt(plaintext, data interface{}, wrappedKey, ciphertext []byte) (err error)
 }
 
+// TaggerInterface provides an API to create a tag, i.e. a piece of information used for authenticating data.
```
Comment (philip-cc) [inline lib/crypto/interfaces.go:29]:
```suggestion
// TaggerInterface provides an API to create a cryptographic tag, i.e. a piece of information used for authenticating data.
```

File: lib/crypto/tagger.go [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+const tagSize = 32
+
+// Tagger implements the TaggerInterface
+type Tagger struct {
+	Key []byte
+}
+
+// NewKMAC256Tagger creates a Tagger which uses KMAC256.
+func NewKMAC256Tagger(key []byte) (Tagger, error) {
+	return Tagger{Key: key}, nil
```
Comment (philip-cc) [inline lib/crypto/tagger.go:31]: We should have a check for the key size here (see `NewAESCryptor`).

File: lib/crypto/tagger_test.go [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+	if err != nil {
+		t.Fatalf("Random failed: %v", err)
+	}
+
+	tagger, err := NewKMAC256Tagger(key)
+	if err != nil {
+		t.Fatalf("NewKMAC256Tagger failed: %v", err)
+	}
+
+	data1, err := rand.GetBytes(uint(1))
```
Comment (philip-cc) [inline lib/crypto/tagger_test.go:33]: I don't think you need the cast to `uint`? (same below)

File: lib/crypto/tagger_test.go [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+	if err != nil {
+		t.Fatalf("Random failed: %v", err)
+	}
+
+	data2, err := rand.GetBytes(uint(2))
+	if err != nil {
+		t.Fatalf("Random failed: %v", err)
+	}
+
+	mac1, err := tagger.Tag(data1)
```
Comment (philip-cc) [inline lib/crypto/tagger_test.go:43]: `s/mac/tag/` just to be consistent.

File: lib/crypto/tagger.go (hunk 5 above)
Comment (AmalieDue) [reply lib/crypto/tagger.go:36]: Fixed

File: lib/crypto/tagger_test.go (hunk 6 above)
Comment (AmalieDue) [reply lib/crypto/tagger_test.go]: Fixed

File: lib/crypto/tagger_test.go (hunk 7 above)
Comment (AmalieDue) [reply lib/crypto/tagger_test.go]: Fixed

File: lib/crypto/tagger.go [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+
+package crypto
+
+import (
+	"bytes"
+	"encoding/gob"
+)
+
+const TagSize = 32
+const KeyLength = 32
```
Comment (philip-cc) [inline lib/crypto/tagger.go:23]: To be consistent and avoid confusion with other similar constants in the `crypto` package. 
```suggestion
const TagLength = 32
const TaggingKeyLength = 32
```

**Review by philip-cc — APPROVED** [review]

**Pull request merged**